explorer/interpreter/stack_space.h

```diff
@@ -0,0 +1,61 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_EXPLORER_INTERPRETER_STACK_SPACE_H_
+#define CARBON_EXPLORER_INTERPRETER_STACK_SPACE_H_
+
+#include <optional>
+
+#include "llvm/ADT/STLFunctionalExtras.h"
+
+namespace Carbon {
+
+namespace Internal {
+
+// Returns true if a new thread should be started for more stack space.
+auto IsStackSpaceNearlyExhausted() -> bool;
+
+// Starts a thread to run the function.
+auto RunWithStackSpaceHelper(llvm::function_ref<void()> fn) -> void;
+
+}  // namespace Internal
+
+// Initializes stack space information for RunWithStackSpace by creating a new
+// thread, and runs the function on the new thread. This only needs to be called
+// once within a given recursion.
+//
+// Usage:
+//   return InitStackSpace<ReturnType>([&]() -> ReturnType {
+//         <function body>
+//       });
+template <typename ReturnType>
+auto InitStackSpace(llvm::function_ref<ReturnType()> fn) -> ReturnType {
```

Comment:
I suggest calling this something like `RunWithNewStack`, to emphasize the fact that this actually does run `fn`.

Reply:
This made me think about the names a bit more -- how about ReserveStackAndRun versus ReserveStackIfExhaustedAndRun? I was thinking that's more explicit about what they do different, and the "AndRun" emphasizes the ordering a little more.

Also, this makes me feel less awkward about sharing implementation a little more.